---
title: Mocking error responses
order: 3
---

To respond with an error, compose a valid HTTP error response by setting the appropriate error HTTP status (`4xx`/`5xx`) on the mocked response. By doing so, you're abiding by the HTTP specification and ensure that your client receives an actual errors response it may encounter in the wild.

<Hint>
  All exceptions originating from request handlers are gracefully handled as 500
  responses by the library, but it's highly recommended{' '}
  <strong>
    to resolve them, as they indicate unexpected behavior in your code
  </strong>
  .
</Hint>

Here's an example of how to respond with a permission error when requesting a `GET /user` without the `Authorization` header set:

```js focusedLines=3-11
rest.get('/user', (req, res, ctx) => {
  // Check when the "Authorization" header is not set.
  if (!req.headers.get('Authorization')) {
    return res(
      // Respond with a "403 Permission denied" status code.
      ctx.status(403),
      // Optionally, provide any additional response data
      // such as headers, error message, etc.
      ctx.json({
        errorMessage: 'Permission denied: not authorized.',
      }),
    )
  }

  // If the "Authorization" header is set, respond with a user.
  return res(
    ctx.json({
      firstName: 'John',
      lastName: 'Maverick',
    }),
  )
})
```

<ResponsePreview
  request={{
    method: 'POST',
    url: '**/login**',
    body: `
{
  "username": "admin"
}
    `,
  }}
  response={{
    statusCode: 403,
    statusText: 'Forbidden',
    body: `
{
  "errorMessage": "User 'admin' not found"
}
    `,
  }}
/>
